### PR TITLE
Allow overriding page title and description in FAQs

### PR DESCRIPTION
**Pull request by dmolineus**
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #161
| Docs PR or issue | -

This PR allows to customize page title and description same as it's allowed for news (#161) and events (c1970b099d6dc071811e3519f80e5c0400d8f6a2)

### CODE REVIEW

**Review by fritzmg — DISMISSED** [review]

**Review by aschempp — DISMISSED** [review]

File: faq-bundle/src/Resources/contao/modules/ModuleFaqReader.php [hunk 1]
```diff
@@ -106,12 +106,29 @@ protected function compile()
 		$this->Template->faq = $objFaq->row();
 
 		// Overwrite the page title and description (see #2853 and #4955)
-		if ($objFaq->question != '')
+		if ($objFaq->pageTitle)
+		{
+			$objPage->pageTitle = strip_tags(StringUtil::stripInsertTags($objFaq->pageTitle));
```
Comment (leofeyer) [inline faq-bundle/src/Resources/contao/modules/ModuleFaqReader.php]:
```suggestion
			$objPage->pageTitle = $objFaq->pageTitle;
```

File: faq-bundle/src/Resources/contao/modules/ModuleFaqReader.php [hunk 2]
```diff
@@ -106,12 +106,29 @@ protected function compile()
...
+		}
+		elseif ($objFaq->question)
 		{
 			$objPage->pageTitle = strip_tags(StringUtil::stripInsertTags($objFaq->question));
+		}
+
+		if ($objFaq->description)
+		{
+			$objPage->description = $this->prepareMetaDescription($objFaq->description);
```
Comment (leofeyer) [inline faq-bundle/src/Resources/contao/modules/ModuleFaqReader.php]:
```suggestion
			$objPage->description = $objFaq->description;
```

File: faq-bundle/src/Resources/contao/models/FaqModel.php [hunk 3]
```diff
@@ -69,6 +75,9 @@
  * @method static Collection|FaqModel[]|FaqModel|null findByAlias($val, $opt=array())
  * @method static Collection|FaqModel[]|FaqModel|null findByAuthor($val, $opt=array())
  * @method static Collection|FaqModel[]|FaqModel|null findByAnswer($val, $opt=array())
+ * @method static Collection|FaqModel[]|FaqModel|null findByPageTitle($val, $opt=array())
+ * @method static Collection|FaqModel[]|FaqModel|null findByRobots($val, $opt=array())
+ * @method static Collection|FaqModel[]|FaqModel|null findByDescription($val, $opt=array())
```
Comment (leofeyer) [inline faq-bundle/src/Resources/contao/models/FaqModel.php:81]: Please also add the `countBy*()` methods.

**Review by leofeyer — CHANGES_REQUESTED** [review]

**Comment by rabauss** [comment]
To have it similar to news + events - it would be nice to have `serpPreview` too, wouldn't it?

**Comment by dmolineus** [comment]
> To have it similar to news + events - it would be nice to have `serpPreview` too, wouldn't it?

Added with 00ecc0c .

File: faq-bundle/src/Resources/contao/models/FaqModel.php (hunk 3 above)
Comment (dmolineus) [reply faq-bundle/src/Resources/contao/models/FaqModel.php:81]: Added with 5872196.

**Comment by dmolineus** [comment]
All suggestions adopted and also added serp preview.

File: faq-bundle/src/Resources/contao/dca/tl_faq.php [hunk 4]
```diff
@@ -486,6 +523,18 @@ public function checkPermission()
 		}
 	}
 
+	public function checkSerpPreview()
+	{
+		$objFaqCategory = FaqCategoryModel::findByPk(CURRENT_ID);
```
Comment (leofeyer) [inline faq-bundle/src/Resources/contao/dca/tl_faq.php]: Not sure if we can/should use the model here. @aschempp Did this not cause problems in some cases?

File: faq-bundle/src/Resources/contao/dca/tl_faq.php [hunk 5]
```diff
@@ -516,6 +565,35 @@ public function generateAlias($varValue, DataContainer $dc)
...
+	 *
+	 * @param FaqModel $objFaq
+	 *
+	 * @return string
+	 */
+	public function getSerpUrl(FaqModel $objFaq)
+	{
+		/** @var FaqCategoryModel $objCategory */
+		$objCategory = $objFaq->getRelated('pid');
+		$jumpTo = (int) $objCategory->jumpTo;
```
Comment (leofeyer) [inline faq-bundle/src/Resources/contao/dca/tl_faq.php:596]: What if `$objCategory` is null?

File: faq-bundle/src/Resources/contao/dca/tl_faq.php [hunk 6]
```diff
@@ -516,6 +565,35 @@ public function generateAlias($varValue, DataContainer $dc)
...
+
+		// A jumpTo page is not mandatory for FAQ categories (see #6226) but required for the FAQ list module
+		if ($jumpTo < 1)
+		{
+			throw new Exception("FAQ categories without redirect page cannot be used in an FAQ list");
```
Comment (leofeyer) [inline faq-bundle/src/Resources/contao/dca/tl_faq.php]:
```suggestion
			throw new Exception('FAQ categories without redirect page cannot be used in an FAQ list');
```

File: faq-bundle/src/Resources/contao/dca/tl_faq.php [hunk 7]
```diff
@@ -516,6 +565,35 @@ public function generateAlias($varValue, DataContainer $dc)
...
+		}
+
+		if (($objTarget = PageModel::findByPk($jumpTo)) !== null)
+		{
+			$strSuffix = StringUtil::ampersand($objTarget->getFrontendUrl(Config::get('useAutoItem') ? '/%s' : '/items/%s'));
+
+			return sprintf(preg_replace('/%(?!s)/', '%%', $strSuffix), ($objFaq->alias ?: $objFaq->id));
+		}
+
+		throw new Exception("FAQ categories without existing redirect page cannot be used in an FAQ list");
```
Comment (leofeyer) [inline faq-bundle/src/Resources/contao/dca/tl_faq.php]:
```suggestion
		throw new Exception('FAQ categories without existing redirect page cannot be used in an FAQ list');
```

**Review by leofeyer — CHANGES_REQUESTED** [review]

**Comment by aschempp** [comment]
@dmolineus can you finish this PR for Contao 4.11? 🙃

**Comment by dmolineus** [comment]
> @dmolineus can you finish this PR for Contao 4.11? upside_down_face

I will, this PR is on my todo list.

File: faq-bundle/src/Resources/contao/dca/tl_faq.php (hunk 5 above)
Comment (dmolineus) [reply faq-bundle/src/Resources/contao/dca/tl_faq.php:596]: This shouldn't never happen in the backend view. In the sense of defensive programming I adjusted it e8307f0639aa9ecdd2eec8cf051307686415c1b3

File: faq-bundle/src/Resources/contao/dca/tl_faq.php (hunk 4 above)
Comment (dmolineus) [reply faq-bundle/src/Resources/contao/dca/tl_faq.php]: Changed with 1e3daf6

**Comment by dmolineus** [comment]
A changed issues mentioned in the feedback and rebased on the current master.

File: faq-bundle/src/Resources/contao/dca/tl_faq.php [hunk 8]
```diff
@@ -520,6 +569,41 @@ public function generateAlias($varValue, DataContainer $dc)
...
+		{
+			throw new Exception('FAQ category does not exist.');
+		}
+
+		$jumpTo = (int) $objCategory->jumpTo;
+
+		// A jumpTo page is not mandatory for FAQ categories (see #6226) but required for the FAQ list module
+		if ($jumpTo < 1)
+		{
+			throw new Exception('FAQ categories without redirect page cannot be used in an FAQ list');
```
Comment (aschempp) [inline faq-bundle/src/Resources/contao/dca/tl_faq.php:601]: should this message be translated (dunno if thats the same for news etc)?

Comment (dmolineus) [reply faq-bundle/src/Resources/contao/dca/tl_faq.php:601]: This message is only shown if any third party extension uses the `getSerpUrl` manually. `tl_faq::checkSerpPreview` will remove the `serpPreview` field for faq categories without a redirect page.

File: faq-bundle/src/Resources/contao/modules/ModuleFaq.php [hunk 9]
```diff
@@ -103,6 +103,11 @@ public function getSearchablePages($arrPages, $intRoot=0, $blnIsSitemap=false)
 				{
 					while ($objItems->next())
 					{
+						if ($blnIsSitemap && $objItems->robots === 'noindex,nofollow')
```
Comment (leofeyer) [inline faq-bundle/src/Resources/contao/modules/ModuleFaq.php:106]: This change seems unrelated. Can you please elaborate?

Comment (dmolineus) [reply faq-bundle/src/Resources/contao/modules/ModuleFaq.php:106]: The description of this PR is misleading. It also contains the robots setting introduced for news and events with https://github.com/contao/contao/pull/1778, see your comment here https://github.com/contao/contao/pull/1778#pullrequestreview-437324426

Comment (leofeyer) [reply faq-bundle/src/Resources/contao/modules/ModuleFaq.php:106]: I see. 👍🏻

**Review by leofeyer — DISMISSED** [review]
Looks good to me. 👍🏻

File: faq-bundle/src/Resources/contao/dca/tl_faq.php [hunk 10]
```diff
@@ -486,6 +523,25 @@ public function checkPermission()
...
+	public function removeMetaFields(DataContainer $dc)
+	{
+		$objFaqCategory = $this->Database
+			->prepare('SELECT c.jumpTo FROM tl_faq_category c LEFT JOIN tl_faq f ON f.pid=c.id WHERE f.id=?')
+			->execute($dc->id);
+
+		if (!$objFaqCategory->jumpTo)
+		{
+			PaletteManipulator::create()
+				->removeField(array('pageTitle', 'robots', 'description', 'serpPreview'), 'meta_legend')
```
Comment (leofeyer) [inline faq-bundle/src/Resources/contao/dca/tl_faq.php:540]: @dmolineus I have adjusted this to remove all meta fields instead of just the `serpPreview` field when the FAQ category does not have a jumpTo page. Does this make sense to you as well?

Comment (dmolineus) [reply faq-bundle/src/Resources/contao/dca/tl_faq.php:540]: LGTM

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @dmolineus.
